cpp/src/decisiontree/batched-levelalgo/builder_base.cuh

```diff
@@ -409,15 +446,19 @@ struct ClsTraits {
       "Builder::computeSplit @builder_base.cuh [batched-levelalgo]");
     auto nbins = b.params.n_bins;
     auto nclasses = b.input.nclasses;
-    auto binSize = nbins * 2 * nclasses;
+    auto binSize = (nbins * 3 + 1) * nclasses;
     auto colBlks = std::min(b.n_blks_for_cols, b.input.nSampledCols - col);
-    dim3 grid(b.n_blks_for_rows, colBlks, batchSize);
     size_t smemSize = sizeof(int) * binSize + sizeof(DataT) * nbins;
     smemSize += sizeof(int);
 
     // Extra room for alignment (see alignPointer in computeSplitClassificationKernel)
     smemSize += 2 * sizeof(DataT) + 1 * sizeof(int);
-
+    int n_blks_for_rows = b.n_blks_for_rows(
+      colBlks,
+      (const void*)
+        computeSplitClassificationKernel<DataT, LabelT, IdxT, TPB_DEFAULT>,
+      TPB_DEFAULT, smemSize);
```

Comment:
This value is computed again in `workspaceSize`. Is there a guarantee that calling `n_blks_for_rows()` would result in a consistent ouput?

Reply:
yes it's value is basically constant, and equals the number of blocks for dimx needed when dimz (parallel nodes) is minimum (which is 1). This happens to be a function of `n_blks_for_cols` and occupancy calculator. As far as occupancy is concerned, the theoretical limiter for now is register count. So ceterus paribus, the output of `n_blks_for_rows()` should also be the same when called from `workspaceSize` and one preceding call of `computeSplitRegressionKernel`